src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -72,106 +73,137 @@ final class TrialService
-              s"No further action will be taken.")
-
-          Future(StatusUpdate.Success)
-        } else {
-          logger.warn(s"Current trial state of the user '${userInfo.userEmail}' is '$currentState'")
-          logger.warn("Checking if enabling is allowed from that state...")
-
-          require(newState.nonEmpty, "Cannot transition to an unspecified state")
-
-          // TODO: Handle invalid initial trial status by
-          //    -adding another case to Trial.StatusUpdate
-          //    -using that case to return a BadRequest from the parent method (i.e. updateUserState())
-          assert(newState.get.isAllowedFrom(currentState),
-            s"Cannot transition from $currentState to $newState")
-
-          // TODO: Test the logic below by adding a mock ThurloeDAO whose saveTrialStatus() returns Failure
-          // Save updates to "userInfo's" trial status, authenticating to Thurloe as "managerInfo" user
-          thurloeDao.saveTrialStatus(userInfo.userSubjectId, managerInfo, newStatus) map {
-            case Success(_) =>
-              logger.warn(s"Updated profile saved as $newState; we are done!")
-              StatusUpdate.Success
-            case Failure(ex) =>
-              StatusUpdate.ServerError(ex.getMessage)
+  private def executeStateTransitions(managerInfo: UserInfo, userEmails: Seq[String],
+                                      createNewState: (WorkbenchUserInfo, Option[UserTrialStatus]) => (UserTrialStatus, Option[Any => Any]),
+                                      handleResult: (Attempt, Option[Any => Any]) => Unit): Future[PerRequestMessage] = {
+    val results: Seq[Future[(String, String)]] = userEmails map { userEmail =>
+      getUserSubjectId(userEmail) flatMap { subId =>
+        val userInfo = WorkbenchUserInfo(subId, userEmail)
+        thurloeDao.getTrialStatus(subId, managerInfo) flatMap { userTrialStatus =>
+          val (newStatus, afterFunction) = createNewState(userInfo, userTrialStatus)
+          if (requiresStateTransition(userTrialStatus, newStatus)) {
+            for {
+              stateResponse <- updateTrialStatus(managerInfo, userInfo, newStatus)
+            } yield {
+              handleResult(stateResponse, afterFunction)
+              (userEmail, StatusUpdate.toName(stateResponse))
+            }
+          } else {
+            logger.warn(
+              s"The user '${userInfo.userEmail}' is already in the trial state of '$newStatus.newState'. " +
+                s"No further action will be taken.")
+            Future.successful((userEmail, StatusUpdate.toName(StatusUpdate.NoChangeRequired)))
           }
         }
       }
-      // TODO: Respond with a more user-friendly error
-      case None => {
-        logger.warn(s"Trial status could not be found for ${userInfo.userEmail}")
+    }
+    Future.sequence(results) map { output => println(output); RequestComplete(output.toMap) }
+  }
```

Comment:
all the passing of anonymous functions - plus the liberal use of `Option` and `Any` - makes this feel pretty incomprehensible to me.

Could you change the signature of `executeStateTransitions` to something like this?
```
 private def executeStateTransitions(managerInfo: UserInfo, userEmails: Seq[String],
                                       statusTransition: UserTrialStatus => UserTrialStatus,
                                       beforeTransitionHandler: (UserInfo, UserTrialStatus, UserTrialStatus) => UserTrialStatus,
                                       afterTransitionHandler: (UserInfo, UserTrialStatus, UserTrialStatus) => UserTrialStatus
                                       ): Future[PerRequestMessage] = {
    val results: Seq[Future[(String, String)]] = userEmails map { userEmail =>
      getUserSubjectId(userEmail) flatMap { subId =>
        val userInfo = WorkbenchUserInfo(subId, userEmail)
        thurloeDao.getTrialStatus(subId, managerInfo) map { currentStatus =>
          val newStatus = statusTransition(currentStatus)
          val (email:String, transitionAttempt:Attempt) = if (requiresStateTransition(currentStatus, newStatus)) {
            for {
              // run custom handler to execute code before performing the transition.
              // before handler can update the state we are trying to transition to.
              statusFromBeforeHandler <- beforeTransitionHandler(managerInfo, currentStatus, newStatus)
              // update thurloe to save the target user's new state
              statusFromTransition <- updateTrialStatus (managerInfo, userInfo, statusFromBeforeHandler)
              // run custom handler to execute code after performing the transition
              // after handler sees the result of what we actually transitioned to
              afterResults <- afterTransitionHandler (managerInfo, currentStatus, statusFromTransition)
            } yield {
              (userEmail, StatusUpdate.Success)
            }
          } else {
            logger.warn(
              s"The user '${userInfo.userEmail}' is already in the trial state of '$newStatus.newState'. " +
                s"No further action will be taken.")
            (userEmail, StatusUpdate.NoChangeRequired)
          }
          (email, StatusUpdate.toName(transitionAttempt))
        }
      }
    }
    Future.sequence(results) map { output => RequestComplete(output.toMap) }
  }
```
if so, the `statusTransition` remains as before this PR, pretty simple; and `beforeTransitionHandler` and `afterTransitionHandler` can each be defined as their own function defs, then passed in. Instead of a single function `enableUsers` which defines anonymous functions inside of it, you could have `enableUsers`, `beforeEnable`, and `afterEnable`.